### PR TITLE
Hypergraph checks

### PR DESCRIPTION
**Pull request by levtelyatnikov**
Overall:
I have reviewed the hypergraph implementation and made necessary modifications to ensure consistent patterns. Additionally, I have updated tutorials and test files to align with these changes. 

Specifically: 
Previously, all the models generated the 'desired' hidden representation, meaning that if a model was required to perform a graph classification task, the pooling operation was executed within the model. Hence, the model output was the pooled signal. 
```
y_pred = model(x_0_init, incidence)
``` 

Now, the models output the last hidden representations. In hypergraph models, the output includes nodes and hyperedges.

```
x_0, x_1 = model(x_0_init, incidence)
``` 

We have introduced a separate, lightweight pooling/readout class that is responsible for providing the desired output for particular downstream tasks. This new pooling/readout class is currently introduced in the tutorials.

```
pooling = readout(task_level="graph")
x_0, x_1 = model(x_0_init, incidence)
y_pred = pooling(x_0, x_1)
```

### CODE REVIEW

**Review by ninamiolane — CHANGES_REQUESTED** [review]
Nice! There seem to be github conflicts though: could you solve them, so that the tests run -- before we review?

File: topomodelx/nn/hypergraph/allset.py [hunk 1]
```diff
@@ -91,15 +87,12 @@ def forward(self, x_0, incidence_1):
...
-        # reversed_edge_index = torch.stack(
-        #     [edge_index[1], edge_index[0]], dim=0)
-
         for layer in self.layers:
-            x_0 = layer(x_0, incidence_1)
-        pooled_x = torch.max(x_0, dim=0)[0]
-        return torch.sigmoid(self.linear(pooled_x))[0]
+            x_0, x_1 = layer(x_0, incidence_1)
+
+        return (x_0, x_1)
```
Comment (ninamiolane) [inline topomodelx/nn/hypergraph/allset.py:98]: No need for (, )

File: topomodelx/nn/hypergraph/allset.py [hunk 2]
```diff
@@ -16,18 +16,16 @@ class AllSet(torch.nn.Module):
...
-    dropout : float
-        Dropout probability.
+    layer_dropout:
+        Dropout probability for the AllSet layer.
     mlp_num_layers : int, default: 2
         Number of layers in the MLP.
-    mlp_activation : torch.nn.Module, default: None
-        Activation function for the MLP.
     mlp_dropout : float, default: 0.0
         Dropout probability for the MLP.
```
Comment (ninamiolane) [inline topomodelx/nn/hypergraph/allset.py:26]: missing parameter type: e.g. `dropout : float` and note the space before : according to numpy's conventions for docstring, which we use here.

File: topomodelx/nn/hypergraph/allset.py [hunk 3]
```diff
@@ -16,18 +16,16 @@ class AllSet(torch.nn.Module):
...
+    mlp_activation:
```
Comment (ninamiolane) [inline topomodelx/nn/hypergraph/allset.py:27]: Ditto.

File: topomodelx/nn/hypergraph/allset_layer.py [hunk 4]
```diff
@@ -109,21 +110,23 @@ def forward(self, x, incidence_1):
...
+        x_1 = self.vertex2edge(x_0, incidence_1.transpose(1, 0))
+        x_1 = F.dropout(x_1, p=self.dropout, training=self.training)
 
-        x = self.edge2vertex(x, incidence_1)
-        x = F.dropout(x, p=self.dropout, training=self.training)
+        x_0 = self.edge2vertex(x_1, incidence_1)
+        x_0 = F.dropout(x_0, p=self.dropout, training=self.training)
 
-        return x
+        return (x_0, x_1)
```
Comment (ninamiolane) [inline topomodelx/nn/hypergraph/allset_layer.py:129]: Ditto

File: topomodelx/nn/hypergraph/allset_transformer.py [hunk 5]
```diff
@@ -16,18 +16,16 @@ class AllSetTransformer(torch.nn.Module):
...
         Dimension of the hidden features.
-    out_channels : int
-        Dimension of the output features.
-    n_layers : int, default: 1
+    n_layers : int, default: 2
         Number of AllSet layers in the network.
     heads : int, default: 4
         Number of attention heads.
-    dropout : float, default=0.2
+    dropout : float
```
Comment (ninamiolane) [inline topomodelx/nn/hypergraph/allset_transformer.py:23]: Ditto

File: topomodelx/nn/hypergraph/allset_transformer.py [hunk 6]
```diff
@@ -16,18 +16,16 @@ class AllSetTransformer(torch.nn.Module):
...
         Dropout probability.
     mlp_num_layers : int, default: 2
         Number of layers in the MLP.
-    mlp_dropout : float, default=0.0
-        Dropout probability for the MLP.
+    mlp_dropout:
```
Comment (ninamiolane) [inline topomodelx/nn/hypergraph/allset_transformer.py:27]: Ditto

File: topomodelx/nn/hypergraph/allset_transformer.py [hunk 7]
```diff
@@ -91,6 +87,6 @@ def forward(self, x_0, incidence_1):
             Output prediction.
         """
         for layer in self.layers:
-            x_0 = layer(x_0, incidence_1)
-        pooled_x = torch.max(x_0, dim=0)[0]
-        return torch.sigmoid(self.linear(pooled_x))[0]
+            x_0, x_1 = layer(x_0, incidence_1)
+
+        return (x_0, x_1)
```
Comment (ninamiolane) [inline topomodelx/nn/hypergraph/allset_transformer.py:92]: Ditto

File: topomodelx/nn/hypergraph/dhgcn.py [hunk 8]
```diff
@@ -50,11 +61,12 @@ def forward(self, x_0):
...
+            Output hyperedge features.
         """
         for layer in self.layers:
-            x_0 = layer(x_0)
-        pooled_x = torch.mean(x_0, dim=0)
-        output = self.linear(pooled_x)
-        return output[0]
+            x_0, x_1 = layer(x_0)
+
+        return (x_0, x_1)
```
Comment (ninamiolane) [inline topomodelx/nn/hypergraph/dhgcn.py:72]: Ditto

File: topomodelx/nn/hypergraph/dhgcn_layer.py [hunk 9]
```diff
@@ -169,25 +169,27 @@ def get_dynamic_topology(self, x_0_features):
         """
         device = x_0_features.device
         n_nodes = x_0_features.size(0)
-        batch = torch.zeros(x_0_features.size(0), dtype=torch.long, device=device)
+        batch = torch.zeros(x_0_features.size(0), dtype=torch.long, device="cpu")
```
Comment (ninamiolane) [inline topomodelx/nn/hypergraph/dhgcn_layer.py:172]: Why hardcoding the device as the cpu? Users might want to process everything on the device of their choice. Consider changing.

File: topomodelx/nn/hypergraph/hmpnn.py [hunk 10]
```diff
@@ -35,32 +35,34 @@ class HMPNN(torch.nn.Module):
...
-        self.to_hidden_linear = torch.nn.Sequential(
-            *[
-                torch.nn.Linear(hidden_features[i], hidden_features[i + 1])
-                for i in range(len(hidden_features) - 1)
-            ]
-        )
+
+        self.linear_node = torch.nn.Linear(in_channels, hidden_channels)
+        self.linear_edge = torch.nn.Linear(in_channels, hidden_channels)
+        # self.to_hidden_linear = torch.nn.Sequential(
```
Comment (ninamiolane) [inline topomodelx/nn/hypergraph/hmpnn.py:48]: No left over comments on main branch: delete (keep it on your own branch if you want)

File: topomodelx/nn/hypergraph/hmpnn.py [hunk 11]
```diff
@@ -35,32 +35,34 @@ class HMPNN(torch.nn.Module):
...
+                    hidden_channels,
                     adjacency_dropout=adjacency_dropout_rate,
                     updating_dropout=regular_dropout_rate,
                 )
-                for _ in range(n_layer)
+                for _ in range(n_layers)
             ]
         )
-        self.to_categories_linear = torch.nn.Linear(hidden_features[-1], num_classes)
+        # self.to_categories_linear = torch.nn.Linear(hidden_features[-1], num_classes)
```
Comment (ninamiolane) [inline topomodelx/nn/hypergraph/hmpnn.py:65]: Ditto

File: topomodelx/nn/hypergraph/hmpnn.py [hunk 12]
```diff
@@ -76,12 +78,15 @@ def forward(self, x_0, x_1, incidence_1):
...
-        x_0 = self.to_hidden_linear(x_0)
-        x_1 = self.to_hidden_linear(x_1)
+        x_0 = self.linear_node(x_0)
+        x_1 = self.linear_edge(x_1)
+
         for layer in self.layers:
             x_0, x_1 = layer(x_0, x_1, incidence_1)
 
-        return self.to_categories_linear(x_0)
+        return (x_0, x_1)
```
Comment (ninamiolane) [inline topomodelx/nn/hypergraph/hmpnn.py:92]: Ditto

File: topomodelx/nn/hypergraph/hmpnn_layer.py [hunk 13]
```diff
@@ -83,18 +83,18 @@ def forward(self, x_source, neighborhood, node_messages):
...
     def forward(self, x_1, m_0):
         return F.sigmoid(self.linear(torch.cat((x_1, m_0), dim=1)))
 
 
 class _DefaultUpdatingFunc(nn.Module):
-    def __init__(self, in_features) -> None:
+    def __init__(self, in_channels) -> None:
         super().__init__()
-        # self.linear = nn.Linear(in_features, in_features)
+        # self.linear = nn.Linear(in_channels, in_channels)
```
Comment (ninamiolane) [inline topomodelx/nn/hypergraph/hmpnn_layer.py:97]: Ditto

File: topomodelx/nn/hypergraph/hnhn_layer.py [hunk 14]
```diff
@@ -196,11 +196,11 @@ def forward(self, x_0, x_1):
...
-            x_1_tp1 += self.bias_0_to_1
+            x_1 += self.bias_0_to_1
         # Compute output hypernode features
-        x_0_tp1 = self.conv_1_to_0(x_1, self.incidence_1)  # edges to nodes
+        x_0 = self.conv_1_to_0(x_1, self.incidence_1)  # edges to nodes
         if self.use_bias:
-            x_0_tp1 += self.bias_1_to_0
-        return torch.sigmoid(x_0_tp1), torch.sigmoid(x_1_tp1)
+            x_0 += self.bias_1_to_0
+        return (torch.relu(x_0), torch.relu(x_1))
```
Comment (ninamiolane) [inline topomodelx/nn/hypergraph/hnhn_layer.py:206]: Ditto

File: topomodelx/nn/hypergraph/unigcn.py [hunk 15]
```diff
@@ -25,36 +25,48 @@ class UniGCN(torch.nn.Module):
...
+        x_1 : torch.Tensor
+            Output hyperedge features.
         """
         for layer in self.layers:
-            x_1 = layer(x_1, incidence_1)
-        pooled_x = torch.max(x_1, dim=0)[0]
-        return torch.sigmoid(self.linear(pooled_x))
+            x_0, x_1 = layer(x_0, incidence_1)
+
+        return (x_0, x_1)
```
Comment (ninamiolane) [inline topomodelx/nn/hypergraph/unigcn.py:72]: Ditto

File: topomodelx/nn/hypergraph/unigcn_layer.py [hunk 16]
```diff
@@ -103,17 +107,19 @@ def forward(self, x_0, incidence_1):
...
         incidence_1_transpose = incidence_1.transpose(1, 0)
-        m_0_1 = self.conv_level1_0_to_1(x_0, incidence_1_transpose)
+        x_1 = self.conv_level1_0_to_1(x_0, incidence_1_transpose)
         if self.bn is not None:
-            m_0_1 = self.bn(m_0_1)
-        m_1_0 = self.conv_level2_1_to_0(m_0_1, incidence_1)
-        return m_1_0
+            x_1 = self.bn(x_1)
+        x_0 = self.conv_level2_1_to_0(x_1, incidence_1)
+        return (x_0, x_1)
```
Comment (ninamiolane) [inline topomodelx/nn/hypergraph/unigcn_layer.py:125]: Ditto

File: topomodelx/nn/hypergraph/unigcnii.py [hunk 17]
```diff
@@ -56,15 +80,19 @@ def forward(self, x_0, incidence_1):
...
         for layer in self.layers:
-            x_0 = layer(x_0, incidence_1, x_0_skip)
+            x_0, x_1 = layer(x_0, incidence_1, x_0_skip)
+            x_0 = self.layer_drop(x_0)
+            x_0 = torch.nn.functional.relu(x_0)
 
-        # linear layer for node classification output
-        # softmax ommited for use of cross-entropy loss
-        return self.linear(x_0)
+        return (x_0, x_1)
```
Comment (ninamiolane) [inline topomodelx/nn/hypergraph/unigcnii.py:98]: Ditto

File: topomodelx/nn/hypergraph/unigcnii.py [hunk 18]
```diff
@@ -29,19 +35,37 @@ class UniGCNII(torch.nn.Module):
...
+    ):
         super().__init__()
         layers = []
-        self.num_features = in_features
-        self.num_classes = num_classes
 
-        for _ in range(num_layers):
+        self.input_drop = torch.nn.Dropout(input_drop)
+        self.layer_drop = torch.nn.Dropout(layer_drop)
+        # Define initial linear layer
```
Comment (ninamiolane) [inline topomodelx/nn/hypergraph/unigcnii.py:53]: Comments are hard to maintain, because people often forget to modify them when they modify code, and that is not caught by unit tests. If you can replace them with variables that are well-named: do it.

for example, here `self.initial_linear_layer` and `self.convolutional_layers` do the job.

File: topomodelx/nn/hypergraph/unigin.py [hunk 19]
```diff
@@ -28,26 +31,30 @@ class UniGIN(torch.nn.Module):
...
+        input_drop=0.2,
+        layer_drop=0.2,
     ):
         super().__init__()
         layers = []
+
+        self.input_drop = torch.nn.Dropout(input_drop)
+        self.layer_drop = torch.nn.Dropout(layer_drop)
+
+        # Define initial linear layer
```
Comment (ninamiolane) [inline topomodelx/nn/hypergraph/unigin.py:47]: Ditto

File: topomodelx/nn/hypergraph/unigin.py [hunk 20]
```diff
@@ -62,11 +69,15 @@ def forward(self, x_0, incidence_1):
...
+        x_0 = self.linear_init(x_0)
         for layer in self.layers:
-            x_0 = layer(x_0, incidence_1)
-        pooled_x_0 = torch.mean(x_0, dim=0)
-        return torch.sigmoid(self.out_decoder(pooled_x_0))
+            x_0, x_1 = layer(x_0, incidence_1)
+            x_0 = self.layer_drop(x_0)
+            x_0 = torch.nn.functional.relu(x_0)
+
+        return (x_0, x_1)
```
Comment (ninamiolane) [inline topomodelx/nn/hypergraph/unigin.py:83]: Ditto

File: topomodelx/nn/hypergraph/unigin_layer.py [hunk 21]
```diff
@@ -89,13 +86,16 @@ def forward(self, x_0, incidence_1):
...
         # First pass fills in features of edges by adding features of constituent nodes
-        m_0_1 = torch.sparse.mm(incidence_1_transpose.float(), x_0)
+        x_1 = torch.sparse.mm(incidence_1_transpose.float(), x_0)
         # Second pass fills in features of nodes by adding features of the incident edges
-        m_1_0 = torch.sparse.mm(incidence_1.float(), m_0_1)
+        m_1_0 = torch.sparse.mm(incidence_1.float(), x_1)
         # Update node features using GIN update equation
-        return self.nn((1 + self.eps) * x_0 + m_1_0)
+        x_0 = self.linear((1 + self.eps) * x_0 + m_1_0)
+        return (x_0, x_1)
```
Comment (ninamiolane) [inline topomodelx/nn/hypergraph/unigin_layer.py:101]: Ditto

File: topomodelx/nn/hypergraph/unisage.py [hunk 22]
```diff
@@ -25,36 +29,59 @@ class UniSAGE(torch.nn.Module):
...
+        # Iterate over layers
         for layer in self.layers:
-            x_1 = layer(x_1, incidence_1)
-        pooled_x = torch.max(x_1, dim=0)[0]
-        return torch.sigmoid(self.linear(pooled_x))
+            x_0, x_1 = layer(x_0, incidence_1)
+            x_0 = self.layer_drop(x_0)
+            x_0 = torch.nn.functional.relu(x_0)
+
+        return (x_0, x_1)
```
Comment (ninamiolane) [inline topomodelx/nn/hypergraph/unisage.py:87]: Ditto

File: topomodelx/nn/hypergraph/unisage_layer.py [hunk 23]
```diff
@@ -106,22 +135,17 @@ def forward(self, x_0, incidence_1):
...
-        # First pass fills in features of edges by adding features of constituent nodes
-        m_0_1 = torch.sparse.mm(incidence_1_transpose.float(), x_0, reduce=self.e_aggr)
-        # Second pass fills in features of nodes by adding features of the incident edges
-        m_1_0 = torch.sparse.mm(incidence_1.float(), m_0_1, reduce=self.v_aggr)
-        return x_0 + m_1_0
+        x_1 = self.vertex2edge(x_0, incidence_1.transpose(1, 0))
+        m_1_0 = self.edge2vertex(x_1, incidence_1)
+        x_0 = x_0 + m_1_0
+
+        return (x_0, x_1)
```
Comment (ninamiolane) [inline topomodelx/nn/hypergraph/unisage_layer.py:151]: Ditto

**Review by ninamiolane — APPROVED** [review]
Perfect! Thank you for this great contribution, and thanks esp. for respecting the naming conventions of our variables. Just minor comments: you can address and then we can merge.

File: topomodelx/nn/hypergraph/dhgcn_layer.py (hunk 9 above)
Comment (levtelyatnikov) [reply topomodelx/nn/hypergraph/dhgcn_layer.py:172]: I have checked and knn and k-means doesn't work with GPU

**Review by ninamiolane — APPROVED** [review]
Awesome, thanks, and thanks for adapting to the new CI!
